compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -502,86 +581,63 @@ mod llvm_enzyme {
             return body;
         }
 
-        let mut exprs = ThinVec::<P<ast::Expr>>::new();
-        if primal_ret {
-            // We have both primal ret and active floats.
-            // primal ret is first, by construction.
-            exprs.push(primal_call);
-        }
-
-        // Now construct default placeholder for each active float.
-        // Is there something nicer than f32::default() and f64::default()?
+        let mut exprs: P<ast::Expr> = primal_call.clone();
         let d_ret_ty = match d_sig.decl.output {
             FnRetTy::Ty(ref ty) => ty.clone(),
             FnRetTy::Default(span) => {
                 panic!("Did not expect Default ret ty: {:?}", span);
             }
         };
-        let mut d_ret_ty = match d_ret_ty.kind.clone() {
-            TyKind::Tup(ref tys) => tys.clone(),
-            TyKind::Path(_, rustc_ast::Path { segments, .. }) => {
-                if let [segment] = &segments[..]
-                    && segment.args.is_none()
-                {
-                    let id = vec![segments[0].ident];
-                    let kind = TyKind::Path(None, ecx.path(span, id));
-                    let ty = P(rustc_ast::Ty { kind, id: ast::DUMMY_NODE_ID, span, tokens: None });
-                    thin_vec![ty]
-                } else {
-                    panic!("Expected tuple or simple path return type");
-                }
-            }
-            _ => {
-                // We messed up construction of d_sig
-                panic!("Did not expect non-tuple ret ty: {:?}", d_ret_ty);
-            }
-        };
-
-        if x.mode.is_fwd() && x.ret_activity == DiffActivity::Dual {
-            assert!(d_ret_ty.len() == 2);
-            // both should be identical, by construction
-            let arg = d_ret_ty[0].kind.is_simple_path().unwrap();
-            let arg2 = d_ret_ty[1].kind.is_simple_path().unwrap();
-            assert!(arg == arg2);
-            let sl: Vec<Symbol> = vec![arg, kw::Default];
-            let tmp = ecx.def_site_path(&sl);
-            let default_call_expr = ecx.expr_path(ecx.path(span, tmp));
-            let default_call_expr = ecx.expr_call(new_decl_span, default_call_expr, thin_vec![]);
-            exprs.push(default_call_expr);
-        } else if x.mode.is_rev() {
-            if primal_ret {
-                // We have extra handling above for the primal ret
-                d_ret_ty = d_ret_ty[1..].to_vec().into();
-            }
 
-            for arg in d_ret_ty.iter() {
-                let arg = arg.kind.is_simple_path().unwrap();
-                let sl: Vec<Symbol> = vec![arg, kw::Default];
-                let tmp = ecx.def_site_path(&sl);
-                let default_call_expr = ecx.expr_path(ecx.path(span, tmp));
+        if x.mode.is_fwd() {
+            // Fwd mode is easy. If the return activity is Const, we support arbitrary types.
+            // Otherwise, we only support a scalar, a pair of scalars, or an array of scalars.
```

Comment:
Please add diagnostics tests for unsupported types

Reply:
You left me thinking for quite a while, but I think we can't find any errors here.

In general `gen_enzyme_decl` (the function called in the line above `gen_enzyme_body`) is responsible for generating the function header based on the original (primal/source) function and the combination of activities. In `gen_enzyme_body` we just try to somehow match the return type, so we should conceptually not really question whether the given function declaration is sensible. That should have been checked in `gen_enzyme_decl`, and for that we have various tests. 

We are still not 100% solid here since we can't look through e.g. alias types, but in the worst case Enzyme will yell at us. In the future we can provide Enzyme with a rust error handler callback, to avoid an ICE.

Reply:
err, actually I can add a ui test for something that doesn't implement default.
I have only added a related test for pretty, I'll add a negative one to UI too.

Reply:
😅 it was mostly the comments that threw me off. Maybe just state that the types were already checked earlier so we don't have any problems here (beyond the Default requirement you just mentioned)

Reply:
Well too late, now we already have another ui test :D 
The error message is helpful, just saying that the user should impl Default for his return type. Let's hope we can move to something like `extern "rust"` soon to delete this whole `gen_enzyme_body` function.